glib/src/object.rs

```diff
@@ -3267,6 +3279,96 @@ wrapper! {
     }
 }
 
+// rustdoc-stripper-ignore-next
+/// A weak reference to an object.
+///
+/// The difference to `WeakRef` is that it isn't thread safe and that it allows to connect a
+/// closure that is called when the set object is disposed.
+#[derive(Debug)]
+pub struct WeakRefLocal<T: ObjectType, F: Fn() + 'static>(
+    *mut (*mut GObject, F),
+    gobject_ffi::GWeakNotify,
+    PhantomData<*mut T>,
+);
+
+impl<T: ObjectType, F: Fn() + 'static> WeakRefLocal<T, F> {
+    /// Create a new empty weak reference.
+    ///
+    /// `upgrade` will always return `None` until an object is set on it.
+    /// The closure will be called when the set object is disposed.
+    pub fn new(f: F) -> WeakRefLocal<T, F> {
+        unsafe extern "C" fn notify_func<F: Fn()>(
+            data: ffi::gpointer,
+            _obj: *mut gobject_ffi::GObject,
+        ) {
+            let data = (data as *mut (*mut GObject, F)).as_mut().unwrap();
+
+            data.0 = ptr::null_mut();
+
+            let f: &F = &*(&data.1 as *const F);
+
+            f();
+        }
+
+        WeakRefLocal(
+            Box::into_raw(Box::new((ptr::null_mut(), f))),
```

Comment:
Why? I think you can also just store the `Box` in your struct :) Also you probably want to make use of `Pin` here because you require that the content of the box is never moved.

Reply:
I didn't yet figure out why how i can use Pin<Box<>> in the place of `*mut` in the struct.